### PR TITLE
REFACTOR: build help pages instead of printing them

### PR DESCRIPTION
**Pull request by amtoine**
# Description
in order to write tests for the `std help` commands, as we currently have in the Rust source base, we need to be able to manipulate the output of the `std help` commands.

however, until now they've all been directly printing the help pages as they go...

this PR tries to build the help pages and return them at the end instead of printing them on the fly :+1: 

> **Note**
> this is quite a rewrite of the `help.nu` module :thinking: 
> i think it might be best to either
> - look at the commits in order
> - test the changes by testing the commands in the REPL and comparing them to their previous `std help` versions

# User-Facing Changes
```
$nothing
```

# Tests + Formatting
- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :black_circle: `toolkit test`
- :black_circle: `toolkit test stdlib`

# After Submitting
```
$nothing
```

### CODE REVIEW

**Comment by amtoine** [comment]
i'm just missing the parameters and flags part of `show-command` :thinking:

**Comment by amtoine** [comment]
i'm a bit off with some newlines and minor format details, but i think it's ready for review :relieved: 

if any reviewer find something off, i'll change it, that should only be small tweaks :wink:

**Comment by amtoine** [comment]
and after a quick test, it looks around as slow as before... (:laughing:)

File: crates/nu-std/lib/help.nu [hunk 1]
```diff
@@ -98,57 +98,55 @@ def "nu-complete list-externs" [] {
...
-    if ($module.env_block? | is-empty) {
-        print $"This module (ansi cyan)does not export(ansi reset) environment."
-    } else {
-        print $"This module (ansi cyan)exports(ansi reset) environment."
-        print (view source $module.env_block)
-    }
+            | str join ', '
+        )"
+        ""
+    ]} else [])
```
Comment (1Kinoti) [inline crates/nu-std/lib/help.nu:134]: i see you using this `if $cond {} else []` syntax, and i think it's not intuitive, (i was surprised that it even works tbh)

i think

```rs
if $cond {} else {}
````

 is better

File: crates/nu-std/lib/help.nu [hunk 2]
```diff
@@ -98,57 +98,55 @@ def "nu-complete list-externs" [] {
...
-
-        let commands_string = (
-            $module.commands
-            | each {|command|
-                $"($command) (char lparen)($module.name) ($command)(char rparen)"
+def build-module-page [module: record] {
+    let usage = (if not ($module.usage? | is-empty) {[
+        $module.usage
+        ""
+    ]} else [])
```
[the remark this answers is not in the document]
Comment (amtoine) [reply crates/nu-std/lib/help.nu:118]: this looks like a sensible suggestion :ok_hand:

[the remark this answers is not in the document]
Comment (amtoine) [reply crates/nu-std/lib/help.nu]: @1Kinoti
is 2686b7675ab6f4c03a9aa4ecd4bb15d9ef4b0c59 any better? :innocent:

[the remark this answers is not in the document]
Comment (1Kinoti) [reply crates/nu-std/lib/help.nu]: yeah, much better!

[the remark this answers is not in the document]
Comment (amtoine) [reply crates/nu-std/lib/help.nu]: cool, thanks for the catch :muscle:

**Comment by amtoine** [comment]
@fdncred 

i ping you for review if you want to have a look before merging :wink: 

i think it looks ok and should not introduce too many differences with the previous version.
and that would open the important door of testing the `std help` module :yum:

**Comment by amtoine** [comment]
> @amtoine What do I have to do to test it? I've tried `use std help` then `std help modules` but that doesn't work any longer.
>
i think jakub broke the prelude in this recent PRs :eyes: 
working on it... :recycle: 

you'd have to run
```
use std
std help modules
```
or
```
use std help
std help
```

**Comment by fdncred** [comment]
Thanks, that helped. Here's my feedback.

- std help = good to go
- std help help = good to go (except we should proably remove/change the help str lpad since that's deprecated)
- std help str lpad = not a fan of the "This command:" section. I'm not sure users care about that information. 
maybe a different format would be more acceptible, so it doesn't take 7 lines? perhaps this?
```
[] scope [x] builtin [x] subcommand [] plugin [] custom [] keyword
```
- std help aliases = good to go
- std help commands = good to go. signatures are different but i'm not sure that matters that much
- std help -f arg = seems like we need to work on usage versus extra_usage. I really don't like seeing so much information in the usage but i don't think there's a way to make extra_usage in custom commands yet.
- std help externs = good to go
- std help externs -f update - coloring is broken on the last line of usage. i wonder if that's a nushell bug?
- std help modules = good to go
- std help operators = good to go
- std help operators -f math = good to go
- std help dirs = i like the old format/look better (highlighted aliases, exported commands don't have command in parens). 
just speculating i wonder if it would be more helpful to show something like?
[[exported_command exported_alias]; [add enter] [drop dexit] [goto g] [next n] [prev p] [show shells]]
- std help xml (or any other std cmd)
not sure about the "Help pages from external command xml"
  - it's not external
  - it's in italic and looks odd


Did i miss anything?

**Comment by amtoine** [comment]
@fdncred 
i'll check that more closely tomorrow :relieved: :wink: 

one question though: **the *bad* points above have been created with this PR or were there before?**
if the latter, i'd rather fix them once we write tests for the `std help` commands, if that's ok with you :blush:

**Comment by fdncred** [comment]
> the bad points above have been created with this PR

This PR. I only checked this PR compared to before this PR.

**Comment by amtoine** [comment]
> This PR. I only checked this PR compared to before this PR.
>
perfect :ok_hand: 

i'm having a look at the points above right now :smirk:

**Comment by amtoine** [comment]
mm i'm a bit confused, all the points you list above look they were there before this PR :confused: 

i even feel they are tracked in the bounty for the most part :open_mouth: 

looks to me we can merge this and address these one at a time next :relieved: 
but yeah i still agree these are issues to be addressed :+1:

**Comment by fdncred** [comment]
I think the problem is that PRs keep getting submitted that don't fix the issues so I end up reviewing PRs over and over again and each PR has the same issues. Maybe going forward we should fix the issues before fixing internal things like build vs print. I really don't want to look at several issues and decide whether I've asked for a fix for this particular thing or not before commenting on it. It makes it really difficult to review PRs that way.

**Comment by amtoine** [comment]
that's true :+1: 

but at the same time, this PR was meant to be a step forward to allow us write tests for the `std help` commands!
until now we've only been eye-balling what does not work.
with this PR, we'll hopefully be able to write Nushell tests for these, as it's done in the Rust side

**Pull request merged**